rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,400 @@
+    auth_service:
+      authentication:
+        connector_name: "my-connector"
+proxy:
+  teleportConfig:
+    proxy_service:
+      acme:
+        enabled: true
+        email: foo@example.com
+```
+
+The following values are core values: users must set them for the chart to work
+properly. They support the happy path. Those values should not be changed by
+the proposal as it would harm backward compatibility and user experience.
+
+- `clusterName`
+- `publicAddr`
+- `chartMode`
+- `aws`
+- `gcp`
+- `enterprise`
+- `operator`
+
+The following values are used to generate Teleport's configuration, we must
+continue to support them for backward compatibility, but using
+`*.teleportConfig` should be preferred.
+
+- `kubeClusterName`
+- `authentication`
+- `authenticationType`
+- `authenticationSecondFactor`
+- `proxyListenerMode`
+- `sessionRecording`
+- `separatePostgresListener`
+- `separateMongoListener`
+- `kubePublicAddr`
+- `mongoPublicAddr`
+- `mysqlPublicAddr`
+- `postgresPublicAddr`
+- `sshPublicAddr`
+- `tunnelPublicAddr`
+- `acme`
+- `acmeEmail`
+- `acmeURI`
+- `log`
+
+#### Chart-specific configuration values
+
+Some values are used to configure the Kubernetes resources deploying Teleport.
+When specified they should apply to both auth and proxy deployments.  Those
+values are:
+
+- `podSecurityPolicy`
+- `labels`
+- `highAvailability`
+- `tls`
+- `image`
+- `enterpriseImage`
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `imagePullPolicy`
+- `initContainers`
+- `postStart`
+- `securityContext`
+- `priorityClassName`
+- `tolerations`
+- `probeTimeoutSeconds`
+- `teleportVersionOverride`
+- `resources`
+
+A few values will have to be treated differently:
+
+- `persistence` will only apply to the `auth` deployment
+- `service` will only apply to the `proxy` service
+- `serviceAccount.name` will apply to the `auth`, the proxy service account
+  name should be the auth one suffixed with `-proxy`
+
+Some users will need to set different values for auth and proxy pods, the
+following values should be also available under `auth` and `proxy`. Those
+specific values should take precedence over the ones at the root.
+
+- `labels`
+- `highAvailability` (except the certManager section)
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `initContainers`
+- `postStart`
+- `tolerations`
+- `teleportVersionOverride`
+- `resources`
+
+### Configuration examples
+
+As this RFD brings numerous value changes and adds several ways of doing the same
+thing, users should be provided full working examples covering various common setups.
+
+Such examples would complete the documentation by demonstrating to users the best
+practices and capabilities of the chart.
+
+Those examples should also be used to lint the chart.
+
+### Update strategy between major versions
+
+Auth pods have to be updated before proxies.  Helm does not support applying
```

Comment:
> Auth pods have to be updated before proxies.

Is this a "must" or a "should"?
Do we have any info regarding this anywhere?
Not really related to this RFD

Reply:
It's a "must" to update auth before proxies and proxies before nodes as Teleport does not provide forward-compatibility guarantees https://goteleport.com/docs/management/operations/upgrading/#upgrade-sequence.

The "auth should be scaled down to 1" part is more complex. In theory, it's a "must", in practice it's a "should", few users are respecting this constraint and we usually try to have backward-compatible storage to be able to roll back if the upgrade fails.